tests/cpp/nodes/test_manipulation.cpp

```diff
@@ -281,6 +281,267 @@ TEST_CASE("ConcatenateNode") {
+                CHECK(std::ranges::equal(ind_ptr->view(state), std::vector{1, 3}));
+                CHECK(std::ranges::equal(v_ptr->view(state), std::vector{-1, -3}));
+
+                CHECK(std::ranges::equal(put_ptr->view(state),
+                                         std::vector{0, -1, 2, -3, 4, 5, 6, 7, 8}));
+                CHECK(std::ranges::equal(put_ptr->mask(state),
+                                         std::vector{0, 1, 0, 1, 0, 0, 0, 0, 0}));
+            }
+
+            THEN("Reverting results in no change") {
+                graph.revert(state, graph.descendants(state, {ind_ptr}));
+                CHECK(std::ranges::equal(put_ptr->view(state),
+                                         std::vector{0, -1, 2, -3, 4, 5, 6, 7, 8}));
+
+                CHECK(std::ranges::equal(put_ptr->mask(state),
+                                         std::vector{0, 1, 0, 1, 0, 0, 0, 0, 0}));
+            }
+        }
+    }
+
+    GIVEN("A 2x3 array of integers, 3 indices with duplicates, and 3 values") {
+        auto graph = Graph();
+
+        auto a_ptr = graph.emplace_node<IntegerNode>(std::initializer_list<ssize_t>{2, 3});
+        auto ind_ptr = graph.emplace_node<IntegerNode>(std::initializer_list<ssize_t>{3}, 0, 5);
+        auto v_ptr = graph.emplace_node<IntegerNode>(std::initializer_list<ssize_t>{3});
+
+        auto put_ptr = graph.emplace_node<PutNode>(a_ptr, ind_ptr, v_ptr);
+
+        graph.emplace_node<ArrayValidationNode>(put_ptr);
+
+        auto state = graph.empty_state();
+        a_ptr->initialize_state(state, {0, 1, 2, 3, 4, 5});
+        ind_ptr->initialize_state(state, {1, 3, 1});
+        v_ptr->initialize_state(state, {10, 20, 30});
+        graph.initialize_state(state);
+
+        // duplicate takes the "most recent", i.e. second, value
+        CHECK(std::ranges::equal(put_ptr->view(state), std::vector{0, 30, 2, 20, 4, 5}));
+        CHECK(std::ranges::equal(put_ptr->mask(state), std::vector{0, 2, 0, 1, 0, 0}));
+
+        WHEN("One of the duplicate values is updated") {
+            v_ptr->set_value(state, 0, 40);
+            graph.propagate(state, graph.descendants(state, {v_ptr}));
+
+            THEN("Commiting keeps the change") {
+                graph.commit(state, graph.descendants(state, {v_ptr}));
+                CHECK(std::ranges::equal(put_ptr->view(state), std::vector{0, 40, 2, 20, 4, 5}));
+                CHECK(std::ranges::equal(put_ptr->mask(state), std::vector{0, 2, 0, 1, 0, 0}));
+            }
+
+            THEN("Reverting results in no change") {
+                graph.revert(state, graph.descendants(state, {v_ptr}));
+                CHECK(std::ranges::equal(put_ptr->view(state), std::vector{0, 30, 2, 20, 4, 5}));
+                CHECK(std::ranges::equal(put_ptr->mask(state), std::vector{0, 2, 0, 1, 0, 0}));
+            }
+        }
+    }
+
+    GIVEN("A length 6 array of integers, and two integer arrays for indexing and values") {
```

Comment:
What clean code, who wrote this section? :stuck_out_tongue:

Reply:
So beautiful, I even [gave credit](https://github.com/dwavesystems/dwave-optimization/pull/201/commits/c1f2015a1db8e6420bd1f9d2e708c7c2dd084c3b)!